### ISSUE TITLE
Discrepancy between template and tutorial snippet

### ISSUE THREAD

**Issue by bittner** [canonical/charmcraft#1116]
### Bug Description

The [`_pebble_layer()` member](https://github.com/canonical/charmcraft/blob/main/charmcraft/templates/init-simple/src/charm.py.j2#L82-L99) in the `init-simple` template seems to be missing code that wraps the Python dictionary into a Pebble `Layer` object. That code, on the other hand, is included in the [Juju Tutorial for Kubernetes](https://juju.is/docs/sdk/create-a-minimal-kubernetes-charm#heading--define-the-charm-initialisation-and-application-services).

It looks like wrapping the dictionary in a `Layer` object has been forgotten (to update?) in the template code.

**EDIT:** Note that I'm a newbie and that the [Juju SDK docs](https://juju.is/docs/sdk/interact-with-pebble#heading--add-a-configuration-layer) mention that it's possible to add a layer both with a plain YAML string or Python dict, and with a Layer object. Hence, the question is, is there [a preferred way](https://pep20.org/#obvious) to do it.

### To Reproduce

1. Go to [GitHub > canonical/charmcraft > charmcraft/templates/init-simple/src/charm.py.j2](https://github.com/canonical/charmcraft/blob/main/charmcraft/templates/init-simple/src/charm.py.j2#L82-L99).
2. Compare with code snippet suggested in [Juju SDK docs > Tutorials > Kubernetes > Create minimal charm](https://juju.is/docs/sdk/create-a-minimal-kubernetes-charm#heading--define-the-charm-initialisation-and-application-services).

### Related

- [Juju SDK docs > How-to Guides > Interact with Pebble](https://juju.is/docs/sdk/interact-with-pebble#heading--add-a-configuration-layer) > Add a configuration layer

**Comment by lengau** [comment]
Thanks for the report! For the question about the preferred ways to add a layer I'd like to get @tonyandrewmeyer or @benhoyt to chime in. I would guess that the `Layer` object is preferred since it'd be the strictest form, but the [ops source](https://github.com/canonical/operator/blob/449793f9867eba11dc2e7e1c28ea3dca1e1da231/ops/pebble.py#L2333) indicates that it's using the strting in the end anyway.

I agree about the templates issue - I'll need to get multiple stakeholders involved to discuss the best way to do that in the templates though.

**Comment by benhoyt** [comment]
Probably if I were building the dict inline right above the `add_layer` call, I'd just use a plain dict. However, given there's a property called `_pebble_layer` I think the K8s tutorial is probably better -- it's return type would be a "proper" `pebble.Layer` object. So my take would be to change the template.

But I'd love to hear if @tonyandrewmeyer has good thoughts either way.

**Comment by tonyandrewmeyer** [comment]
Yeah, given that `add_layer` will immediately convert a `LayerDict` to a `Layer` it doesn't really make much difference.

I agree with both of @benhoyt's points. In addition:

If I was writing from scratch, I would generally use a `Layer`, because I'd start writing:

```python
def _pebble_layer(self):
    return ops.pebble.Layer({"
```

And at that point I'd get autocomplete for the fields. However, if you do this:

```python
def _pebble_layer(self) -> ops.pebble.LayerDict:
    return {"
```

Then you get the autocomplete too, so it's not a strong advantage (I generally prefer the implicit return types).

I prefer it for testing the layer property, too, e.g. this:

```python
assert layer.services["x"].command == 
```

Rather than

```python
assert layer["services"]["x"]["command"] ==
```

And also when testing events that need the layer to be present for some reason, e.g.:

```python
layer = ops.pebble.Layer({...})
container = Container(layers={layer})
...
assert layer in state.get_container(container.name).layers
```

Although this is also not much of an issue since `Layer`s can be compared to `LayerDict`s, but it feels cleaner to be comparing the same types.

If I didn't know how it was implemented, I think it would also 'feel' stronger - I'd imagine that there could be additional value checking when creating the object and wouldn't know that the object creation happens anyway.

I would definitely _not_ pass a YAML string, unless I was sourcing that from somewhere else (like a file). I think that's in the how-to guide because it's directly stating what's in the reference docs rather than because it's an equally common choice.

**Comment by dwilding** [comment]
Sorry, I was comparing `Layer` vs `LayerDict` as the return type. (Thanks for giving your thoughts on the other points - good to know 🙂)

Looking again with fresh Monday eyes, I think `_get_pebble_layer` feels fussy because I have to remember to include `LayerDict` to satisfy the type checker. I know it's not a huge deal, but my memory from originally working through the tutorial is something like "the `Layer` way is hard to get right".

We must have discussed this somewhere before - but would there be disadvantages to making `Layer` also accept a plain dict?

**Comment by dwilding** [comment]
Ah yes, thanks Dima, this was done in https://github.com/canonical/operator/pull/2003. @lengau this can be closed (I don't have permission)

**Comment by mr-cal** [comment]
Closed per @dwilding's rationale
